### PR TITLE
Feature/add featureset as variant snowflake

### PR DESCRIPTION
**Pull request by swatiarora0208**
**What this PR does / why we need it**:
1. Implements the prepareWrite and write feature to snowflake using jdbc for batch store
2. Creates the snowflake table for feature row if not already exists
3. insert feature row as variant type column in the table

Fixes #

**Does this PR introduce a user-facing change?**:

### CODE REVIEW

File: storage/connectors/jdbc/src/main/java/feast/storage/connectors/jdbc/snowflake/SnowflakeTemplater.java [hunk 1]
```diff
@@ -47,79 +54,133 @@ public String getTableCreationSql(FeatureSetProto.FeatureSetSpec featureSetSpec)
...
+
+    StringJoiner columnsSql = new StringJoiner(",");
+    StringJoiner valueSql = new StringJoiner(",");
 
-    if (requiredColumns.size() == 0) {
-      log.info(
-          String.format("All columns already exist for table %s, no update necessary.", tableName));
-      return tableMigrationSql;
+    String tableName = JdbcTemplater.getTableNameFromFeatureSet(element.getFeatureSet());
+    //    Map<String, String> requiredColumns = getRequiredColumns(featureSetSpec);
```
Comment (dr3s) [inline storage/connectors/jdbc/src/main/java/feast/storage/connectors/jdbc/snowflake/SnowflakeTemplater.java]: delete

File: storage/connectors/jdbc/pom.xml [hunk 2]
```diff
@@ -126,9 +125,13 @@
             <groupId>org.mockito</groupId>
             <artifactId>mockito-core</artifactId>
         </dependency>
+  		 <dependency>
```
Comment (dr3s) [inline storage/connectors/jdbc/pom.xml]: formatting

File: storage/connectors/jdbc/src/main/java/feast/storage/connectors/jdbc/common/JdbcTemplater.java [hunk 3]
```diff
@@ -32,12 +33,15 @@ static String getTableName(FeatureSetProto.FeatureSetSpec featureSetSpec) {
         .replaceAll("-", "_");
   }
 
-  String getTableCreationSql(FeatureSetProto.FeatureSetSpec featureSetSpec);
+  static String getTableNameFromFeatureSet(String featureSet) {
+    System.out.println("getTableNameFromFeatureSet--" + featureSet);
```
Comment (dr3s) [inline storage/connectors/jdbc/src/main/java/feast/storage/connectors/jdbc/common/JdbcTemplater.java]: is this the convention or is there a proper logger?

File: storage/connectors/jdbc/src/main/java/feast/storage/connectors/jdbc/snowflake/SnowflakeTemplater.java [hunk 4]
```diff
@@ -31,8 +38,8 @@
   @Override
   public String getTableCreationSql(FeatureSetProto.FeatureSetSpec featureSetSpec) {
     StringJoiner columnsAndTypesSQL = new StringJoiner(", ");
-    Map<String, String> requiredColumns = getRequiredColumns(featureSetSpec);
-
+    //    Map<String, String> requiredColumns = getRequiredColumns(featureSetSpec);
```
Comment (dr3s) [inline storage/connectors/jdbc/src/main/java/feast/storage/connectors/jdbc/snowflake/SnowflakeTemplater.java]: delete

File: storage/connectors/jdbc/src/main/java/feast/storage/connectors/jdbc/snowflake/SnowflakeTemplater.java [hunk 5]
```diff
@@ -47,79 +54,133 @@ public String getTableCreationSql(FeatureSetProto.FeatureSetSpec featureSetSpec)
...
-  public String getFeatureRowInsertSql(FeatureSetProto.FeatureSetSpec featureSetSpec) {
-    StringJoiner columnsSql = new StringJoiner(",");
-    StringJoiner valueSql = new StringJoiner(",");
+    // Set ingestion Id
 
-    Map<String, String> requiredColumns = getRequiredColumns(featureSetSpec);
+    if (element.getIngestionId() == "") {
+      valueSql.add("NULL");
+    } else {
+      valueSql.add("'" + element.getIngestionId() + "'");
```
Comment (dr3s) [inline storage/connectors/jdbc/src/main/java/feast/storage/connectors/jdbc/snowflake/SnowflakeTemplater.java]: use parameter binding not string concatenation

File: storage/connectors/jdbc/src/main/java/feast/storage/connectors/jdbc/snowflake/SnowflakeTemplater.java [hunk 6]
```diff
@@ -47,79 +54,133 @@ public String getTableCreationSql(FeatureSetProto.FeatureSetSpec featureSetSpec)
...
+    for (String row : fieldMap.keySet()) {
+      setFeatureValue(json_variant, row, fieldMap.get(row));
     }
 
-    tableMigrationSql =
-        tableMigrationSql + String.format("ALTER TABLE %s %s;", tableName, addColumnSql);
-    log.debug(tableMigrationSql);
-    return tableMigrationSql;
-  }
+    valueSql.add("parse_json('" + json_variant.toString() + "')");
```
Comment (dr3s) [inline storage/connectors/jdbc/src/main/java/feast/storage/connectors/jdbc/snowflake/SnowflakeTemplater.java]: use parameter binding not string concatenation

File: storage/connectors/jdbc/src/main/java/feast/storage/connectors/jdbc/snowflake/SnowflakeTemplater.java [hunk 7]
```diff
@@ -47,79 +54,133 @@ public String getTableCreationSql(FeatureSetProto.FeatureSetSpec featureSetSpec)
...
-    StringJoiner addColumnSql = new StringJoiner(", ");
-    // Filter required columns down to only the ones we need to add
-    for (String requiredColumn : requiredColumns.keySet()) {
-      String requiredColumnType = requiredColumns.get(requiredColumn);
-      addColumnSql.add(String.format("ADD COLUMN %s %s", requiredColumn, requiredColumnType));
+    valueSql.add("'" + Timestamp.from(eventTsInstant) + "'");
+
+    // Set created_timestamp
+
+    valueSql.add("'" + System.currentTimeMillis() + "'");
```
Comment (dr3s) [inline storage/connectors/jdbc/src/main/java/feast/storage/connectors/jdbc/snowflake/SnowflakeTemplater.java]: use parameter binding not string concatenation

File: storage/connectors/jdbc/src/main/java/feast/storage/connectors/jdbc/snowflake/SnowflakeTemplater.java [hunk 8]
```diff
@@ -47,79 +54,133 @@ public String getTableCreationSql(FeatureSetProto.FeatureSetSpec featureSetSpec)
...
+    Instant eventTsInstant =
+        Instant.ofEpochSecond(element.getEventTimestamp().getSeconds())
+            .plusNanos(element.getEventTimestamp().getNanos());
 
-    StringJoiner addColumnSql = new StringJoiner(", ");
-    // Filter required columns down to only the ones we need to add
-    for (String requiredColumn : requiredColumns.keySet()) {
-      String requiredColumnType = requiredColumns.get(requiredColumn);
-      addColumnSql.add(String.format("ADD COLUMN %s %s", requiredColumn, requiredColumnType));
+    valueSql.add("'" + Timestamp.from(eventTsInstant) + "'");
```
Comment (dr3s) [inline storage/connectors/jdbc/src/main/java/feast/storage/connectors/jdbc/snowflake/SnowflakeTemplater.java]: use parameter binding not string concatenation

File: storage/connectors/jdbc/src/main/java/feast/storage/connectors/jdbc/snowflake/SnowflakeTemplater.java [hunk 9]
```diff
@@ -47,79 +54,133 @@ public String getTableCreationSql(FeatureSetProto.FeatureSetSpec featureSetSpec)
...
+      valueSql.add("NULL");
+    } else {
+      valueSql.add("'" + element.getIngestionId() + "'");
+    }
 
-    for (String column : requiredColumns.keySet()) {
-      columnsSql.add(column);
-      valueSql.add("?");
+    // Set job Name
+    valueSql.add("'" + jobName + "'");
```
Comment (dr3s) [inline storage/connectors/jdbc/src/main/java/feast/storage/connectors/jdbc/snowflake/SnowflakeTemplater.java]: use parameter binding not string concatenation

File: storage/connectors/jdbc/src/main/java/feast/storage/connectors/jdbc/writer/InsertFeatureSetToTable.java [hunk 10]
```diff
@@ -0,0 +1,112 @@
...
+      props.put("password", this.jdbcConfig.getPassword());
+      props.put("db", this.jdbcConfig.getDatabase());
+      props.put("schema", this.jdbcConfig.getSchema());
+      props.put("tracing", "ALL");
+      props.put("warehouse", this.jdbcConfig.getWarehouse());
+      Connection conn = DriverManager.getConnection(this.jdbcConfig.getUrl(), props);
+
+      DatabaseMetaData meta = conn.getMetaData();
+      //      System.out.println("check if table exists::::"+meta.getTables(null, null,
+      // tableName.toUpperCase(), null).next());
```
Comment (dr3s) [inline storage/connectors/jdbc/src/main/java/feast/storage/connectors/jdbc/writer/InsertFeatureSetToTable.java]: delete

File: storage/connectors/jdbc/src/main/java/feast/storage/connectors/jdbc/writer/InsertFeatureSetToTable.java [hunk 11]
```diff
@@ -0,0 +1,112 @@
...
+      if (!meta.getTables(null, null, tableName.toUpperCase(), null).next()) {
+
+        System.out.println("if table does not exists //");
```
Comment (dr3s) [inline storage/connectors/jdbc/src/main/java/feast/storage/connectors/jdbc/writer/InsertFeatureSetToTable.java]: Use logger

File: storage/connectors/jdbc/src/main/java/feast/storage/connectors/jdbc/writer/InsertFeatureSetToTable.java [hunk 12]
```diff
@@ -0,0 +1,112 @@
...
+        Statement stmt = conn.createStatement();
+        stmt.execute(createSqlTableCreationQuery);
+      }
+      String insertQuery = this.jdbcTemplater.getFeatureRowInsertSql(element, this.jobName);
+      Statement statement = conn.createStatement();
+      statement.executeQuery(insertQuery);
+      conn.close();
+
+      //
+      // snowflakeFeatureSinkObj.prepareWrite(element.apply("CreateSchema",Create.of(specMap)));
```
Comment (dr3s) [inline storage/connectors/jdbc/src/main/java/feast/storage/connectors/jdbc/writer/InsertFeatureSetToTable.java]: delete

File: storage/connectors/jdbc/src/test/java/feast/storage/connectors/jdbc/writer/JdbcSnowflakeFeatureSinkTest.java [hunk 13]
```diff
@@ -156,20 +126,17 @@ public void shouldWriteToSnowflake() throws SQLException {
             FeatureRow.newBuilder()
```
Comment (dr3s) [inline storage/connectors/jdbc/src/test/java/feast/storage/connectors/jdbc/writer/JdbcSnowflakeFeatureSinkTest.java:162]: I think there needs to be more extensive testing of both success and failure cases.  One of the big ones we need to check is how this job responds when features change after it has been started.  I'd like to see some thought put into others as well.

File: storage/connectors/jdbc/src/main/java/feast/storage/connectors/jdbc/common/JdbcTemplater.java (hunk 3 above)
Comment (zhangchi1) [reply storage/connectors/jdbc/src/main/java/feast/storage/connectors/jdbc/common/JdbcTemplater.java]: I think this is just for our testing purpose, correct me if I'm wrong @swatiarora0208

Comment (swatiarora0208) [reply storage/connectors/jdbc/src/main/java/feast/storage/connectors/jdbc/common/JdbcTemplater.java]: it is not the logger.. removed it

File: storage/connectors/jdbc/src/main/java/feast/storage/connectors/jdbc/snowflake/SnowflakeTemplater.java (hunk 5 above)
Comment (swatiarora0208) [reply storage/connectors/jdbc/src/main/java/feast/storage/connectors/jdbc/snowflake/SnowflakeTemplater.java]: @dr3s did you mean preparedStatement.setString() method?

Comment (dr3s) [reply storage/connectors/jdbc/src/main/java/feast/storage/connectors/jdbc/snowflake/SnowflakeTemplater.java]: yes

**Review by dr3s — APPROVED** [review]

File: protos/feast/core/Store.proto [hunk 14]
```diff
@@ -147,6 +147,7 @@ message Store {
     string database = 7;
     string schema = 8;
     string warehouse = 9;
+    string tablename = 10;
```
Comment (zhangchi1) [inline protos/feast/core/Store.proto]: I'm using `table` as the key of the config in the retriever, at [pos1](https://github.com/Cimpress-MCP/feast/blob/11550ccbf989e4da74897df37cffd41fa49afb79/storage/connectors/jdbc/src/main/java/feast/storage/connectors/jdbc/retriever/SnowflakeQueryTemplater.java#L41), and at [pos2](https://github.com/Cimpress-MCP/feast/blob/11550ccbf989e4da74897df37cffd41fa49afb79/storage/connectors/jdbc/src/test/java/feast/storage/connectors/jdbc/retriever/SnowflakeHistoricalRetrieverJSONColTest.java#L62). So maybe you can update them to `tableName`, when you do rebase ~

**Pull request merged**